dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/Compression.kt

```diff
@@ -0,0 +1,105 @@
+package org.jetbrains.kotlinx.dataframe.io
+
+import java.io.File
+import java.io.InputStream
+import java.net.URL
+import java.nio.file.Path
+import java.util.zip.GZIPInputStream
+import java.util.zip.InflaterInputStream
+import java.util.zip.ZipInputStream
+
+/**
+ * Compression algorithm to use when reading csv files.
+ * We support [GZIP][Compression.Gzip] and [ZIP][Compression.Zip] compression out of the box.
+ *
+ * Custom decompression algorithms can be added by creating an instance of [Custom].
+ *
+ * @param wrapStream function that wraps any [InputStream] into a decompressing [InflaterInputStream] stream
+ */
+public sealed class Compression<I : InputStream>(public open val wrapStream: (InputStream) -> I) {
+
+    /** Can be overridden to perform some actions before reading from the input stream. */
+    public open fun doFirst(inputStream: I) {}
+
+    /**
+     * Can be overridden to perform some actions after reading from the input stream.
+     * Remember to close the stream if you override this function.
+     */
+    public open fun doFinally(inputStream: I) {
+        inputStream.close()
+    }
+
+    /**
+     * For .gz / GZIP files.
+     */
+    public data object Gzip : Compression<GZIPInputStream>(wrapStream = ::GZIPInputStream)
+
+    /**
+     * For .zip / ZIP files.
+     */
+    public data object Zip : Compression<ZipInputStream>(wrapStream = ::ZipInputStream) {
+
+        override fun doFirst(inputStream: ZipInputStream) {
+            // Make sure to call nextEntry once to prepare the stream
+            if (inputStream.nextEntry == null) error("No entries in zip file")
+        }
+
+        override fun doFinally(inputStream: ZipInputStream) {
+            // Check we don't have more than one entry in the zip file
+            if (inputStream.nextEntry != null) {
+                inputStream.close()
+                throw IllegalArgumentException("Zip file contains more than one entry")
+            }
+            inputStream.close()
+        }
+    }
+
+    /**
+     * No compression.
+     */
+    public data object None : Compression<InputStream>(wrapStream = { it })
```

Comment:
Seems like "None" is actually kind of "Detect", no? At least according to kdocs it seems like it is

Reply:
no, it's actually "none". If a user supplies `compression = Compression.None`, the inputStream will be supplied to deephaven-csv directly.

If a user does _not_ supply a compression argument, the default will be "`Compression.None`, unless detected otherwise from the input file or url." So `compressionStateOf()` will be called (if possible) and the right one will be chosen based on the file extension, defaulting to `None` if that fails.

Hope that clarifies it :) If not, please tell me which part can be written clearer